exercises/practice/zebra-puzzle/.meta/src/ZebraPuzzle.example.elm

```diff
@@ -0,0 +1,321 @@
+    ]
+
+
+neighbors : Number -> List Number
+neighbors number =
+    case number of
+        One ->
+            [ Two ]
+
+        Two ->
+            [ One, Three ]
+
+        Three ->
+            [ Two, Four ]
+
+        Four ->
+            [ Three, Five ]
+
+        Five ->
+            [ Four ]
+
+
+rightNeighbor : Number -> List Number
+rightNeighbor number =
+    case number of
+        One ->
+            [ Two ]
+
+        Two ->
+            [ Three ]
+
+        Three ->
+            [ Four ]
+
+        Four ->
+            [ Five ]
+
+        Five ->
+            []
+
+
+leftNeighbor : Number -> List Number
+leftNeighbor number =
+    case number of
+        One ->
+            []
+
+        Two ->
+            [ One ]
+
+        Three ->
+            [ Two ]
+
+        Four ->
+            [ Three ]
+
+        Five ->
+            [ Four ]
+
+
+neighboringHouseConstraint : (Number -> List Number) -> (House -> a) -> a -> (House -> b) -> b -> List House -> House -> Bool
+neighboringHouseConstraint getNeighbors field1 val1 field2 val2 houses house =
+    let
+        numbersOfHousesWithField1 =
+            List.filter (\h -> field1 h == val1) houses
+                |> List.map .number
+                |> List.concatMap getNeighbors
+    in
+    field2 house /= val2 || List.member (.number house) numbersOfHousesWithField1
+
+
+applyneighboringHouseConstraint : List House -> List House
+applyneighboringHouseConstraint houses =
+    List.filter (\house -> List.all (\constraint -> constraint houses house) neighboringHouseConstraints) houses
+
+
+findNewSameHouseConstraints : List House -> List (House -> Bool)
+findNewSameHouseConstraints houses =
+    List.concatMap
+        (\house ->
+            List.filterMap identity
+                [ newConstraint .nationality house houses
+                , newConstraint .color house houses
+                , newConstraint .drink house houses
+                , newConstraint .animal house houses
+                , newConstraint .cigaretteBrand house houses
+                ]
+        )
+        houses
+
+
+newConstraint : (House -> a) -> House -> List House -> Maybe (House -> Bool)
+newConstraint field house houses =
+    let
+        typesOfFieldInHouse =
+            List.filterMap
+                (\h ->
+                    if .number h == .number house then
+                        Just (field h)
+
+                    else
+                        Nothing
+                )
+                houses
+
+        houseOnlyHasOneTypeOfField =
+            List.length (unique typesOfFieldInHouse) == 1
+
+        typeOfFieldIsOnlyFoundInHouse =
+            List.all (\h -> field h /= field house || .number h == .number house) houses
+    in
+    if houseOnlyHasOneTypeOfField || typeOfFieldIsOnlyFoundInHouse then
+        Just (sameHouseConstraint field (field house) .number (.number house))
+
+    else
+        Nothing
+
+
+unique : List a -> List a
```

Comment:
I think using a Set here is more usual in Elm

Reply:
Of course, but `Set` won't take a custom types, and I didn't feel like defining some `toComparable` and `fromComparable` for all 5 types of custom types I use this function on. I really wish `Set` and `Dict` could handle any type out of the box.